File: packages/client/src/app/extensions.ts
import {
	NotificationType,
	RequestType,
	TextDocumentIdentifier,
} from "vscode-languageclient/node";

export interface UnreadDependencyParams {
	dependency: string;
	dependant: string;
}

export interface DebugDocumentParams {
	textDocument: TextDocumentIdentifier;
}

// prettier-ignore
namespace ext {
	/* eslint-disable prefer-let/prefer-let */
	export const UNREAD_DEPENDENCY =
		new NotificationType<UnreadDependencyParams>("wgsl/unreadDependency");
	export const DEBUG_AST =
		new RequestType<DebugDocumentParams, string, void>("wgsl/debugAst");
	export const DEBUG_TOKENS =
		new RequestType<DebugDocumentParams, string, void>("wgsl/debugTokens");
}

export default ext;

File: packages/server/src/extensions.rs
use lsp_types::{notification::Notification, request::Request, TextDocumentIdentifier, Url};

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UnreadDependencyParams {
	pub dependency: Url,
	pub dependant: Url,
}

pub enum UnreadDependency {}
impl Notification for UnreadDependency {
	type Params = UnreadDependencyParams;
	const METHOD: &'static str = "wgsl/unreadDependency";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct DebugDocumentParams {
	pub text_document: TextDocumentIdentifier,
}

pub enum DebugAst {}
impl Request for DebugAst {
	type Params = DebugDocumentParams;
	type Result = String;
	const METHOD: &'static str = "wgsl/debugAst";
}

pub enum DebugTokens {}
impl Request for DebugTokens {
	type Params = DebugDocumentParams;
	type Result = String;
	const METHOD: &'static str = "wgsl/debugTokens";
}
